### ISSUE TITLE
java.lang.OutOfMemoryError: Java heap space

### ISSUE THREAD

**Issue by montague** [adamtornhill/code-maat#2]
I'm following along and tried to run the analysis tool on some logs generated from a ~3 year old project. there are about 81K lines in the file i'm trying to analyze -- here's the command and output:
```
$ ~/opt/ixmaat0.8.2/maat -l ignore_me/commit_log.txt -c git -a summary
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at clojure.lang.PersistentHashMap$ArrayNode.assoc(PersistentHashMap.java:368)
	at clojure.lang.PersistentHashMap$ArrayNode.assoc(PersistentHashMap.java:365)
	at clojure.lang.PersistentHashMap.assoc(PersistentHashMap.java:142)
	at clojure.lang.PersistentHashMap.assoc(PersistentHashMap.java:28)
	at clojure.lang.RT.assoc(RT.java:702)
	at clojure.core$assoc.invoke(core.clj:187)
	at clojure.lang.Atom.swap(Atom.java:65)
	at clojure.core$swap_BANG_.invoke(core.clj:2234)
	at instaparse.gll$node_get.invoke(gll.clj:221)
	at instaparse.gll$push_listener.invoke(gll.clj:267)
	at instaparse.gll$CatListener$fn__672.invoke(gll.clj:397)
	at instaparse.gll$push_message$f__602.invoke(gll.clj:173)
	at instaparse.gll$step.invoke(gll.clj:328)
	at instaparse.gll$run.invoke(gll.clj:344)
	at instaparse.gll$run.invoke(gll.clj:332)
	at instaparse.gll$parse.invoke(gll.clj:758)
	at instaparse.core$parse.doInvoke(core.clj:83)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at code_maat.parsers.hiccup_based_parser$as_grammar_map.invoke(hiccup_based_parser.clj:23)
	at code_maat.parsers.hiccup_based_parser$parse_log.invoke(hiccup_based_parser.clj:90)
	at code_maat.parsers.git$parse_log.invoke(git.clj:62)
	at code_maat.app.app$git__GT_modifications$fn__8177.invoke(app.clj:111)
	at code_maat.app.app$run_parser_in_error_handling_context.invoke(app.clj:80)
	at code_maat.app.app$git__GT_modifications.invoke(app.clj:109)
	at code_maat.app.app$parse_commits_to_dataset.invoke(app.clj:168)
	at code_maat.app.app$run.invoke(app.clj:181)
	at code_maat.cmd_line$_main.doInvoke(cmd_line.clj:63)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at code_maat.cmd_line.main(Unknown Source)
```

**Comment by adamtornhill** [comment]
Code Maat is quite memory hungry. The problem is in the parsers for Git and Mercurial (Subversion uses a different strategy and performs better). Those parsers are based on Instaparse. While Instaparse is a great tool, Instaparse does use a lot of memory (see the documentation [here](https://github.com/Engelberg/instaparse/blob/master/docs/Performance.md)).

I'm gonna experiment a bit with the algorithms. One possibility is to split the log into smaller chunks and parse those individually. Perhaps the grammar can be improved as well.

Contributions are welcome!

**Comment by adamtornhill** [comment]
I re-wrote the parsers to work on lazy sequences built from tokenized input. This way, we parse the entries in the version-control files one by one. The prototype shows a large improvement in memory consumption.
I need to update all test programs, re-implement the encoding support and modify the Git grammar to handle merges.

**Comment by montague** [comment]
that sounds great, thanks! i've never written clojure, but i'm happy to test builds if that would help.

**Comment by adamtornhill** [comment]
Thanks - I appreciate that! Could you please e-mail me and I'll send you a build with some details once I get everything together? You can reach me at: adam at adamtornhill dot com

**Comment by adamtornhill** [comment]
This issue has been resolved in version 0.8.3. I re-wrote the parsers to work on pre-processed chunks of data from the log file.
Thanks to montague for helping out with the tests!

**Comment by montague** [comment]
:+1: for open source!
